assets/js/googlesitekit/modules/datastore/modules.js

```diff
@@ -154,6 +167,58 @@ const baseActions = {
 
 		return { response, error };
 	},
+
+	/**
+	 * Registers a module
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {string}          slug                         Module slug.
+	 * @param {Object}          settings                     Module settings.
+	 * @param {string}          [settings.name]              Optional. Module name. Default is the slug.
+	 * @param {string}          [settings.description]       Optional. Module description. Default empty string.
+	 * @param {string}          [settings.icon]              Optional. Module icon. Default empty string.
+	 * @param {number}          [settings.order]             Optional. Numeric indicator for module order. Default 10.
+	 * @param {string}          [settings.homepage]          Optional. Module homepage URL. Default empty string.
+	 * @param {React.Component} [settings.settingsComponent] React component to render the settings panel. Default is the DefaultModuleSettings component.
+	 * @return {Object} Redux-style action.
+	 */
+	*registerModule( slug, { settingsComponent = DefaultModuleSettings, ...settings } = {} ) {
+		invariant( slug, 'module slug is required' );
+
+		const registry = yield commonActions.getRegistry();
+		let registryKey = yield registry.select( CORE_SITE_STORE_NAME ).getRegistryKey();
+		if ( registryKey === undefined ) {
+			registryKey = Object.keys( ModuleComponents ).length + 1;
```

Comment:
Shoot, I just realised an issue with this approach: we're declaring a new key based on the number of `ModuleComponents` registered—in this case `Object.keys( ModuleComponents ).length + 1;` might equal `1`, so we'll set this registry's key as `1`, but what if multiple registries have been registered with widgets first and there are is already a `1` and `2` key in the registry?

Either we need to use UUIDs here (not a bad idea really) or we need to also check that the registry key we create from this isn't already in use. Otherwise we could run into very edge-case issues with registries sharing a key.

Reply:
(I'd recommend us creating UUIDs for the registry keys rather than our naive integer approach.)